site/app/Services/Clone/CloneCardService.php

```diff
@@ -0,0 +1,219 @@
+<?php
+
+namespace App\Services\Clone;
+
+use App\Card;
+use App\Course;
+use App\Enums\FileStatus;
+use App\Enums\StateType;
+use App\Exceptions\CloneException;
+use App\Folder;
+use Illuminate\Support\Facades\DB;
+use InvalidArgumentException;
+
+class CloneCardService
+{
+    private Card $card;
+
+    public function __construct(Card $card)
+    {
+        $this->card = $card;
+    }
+
+    /**
+     * Check if the card can be cloned.
+     *
+     * Same params than CloneService::clone().
+     *
+     * @throws CloneException If the card cannot be cloned.
+     */
+    public function checkClone(
+        Folder $destFolder = null,
+        Course $destCourse = null,
+    ): void {
+        $cloneInAnotherSpace = (false
+            || $destFolder && $destFolder->course->id !== $this->card->course->id
+            || $destCourse && $destCourse->id !== $this->card->course->id
+        );
+        if ($cloneInAnotherSpace) {
+            // Check that the source file is ready (not transcoding).
+            if ($this->card->file && $this->card->file->status !== FileStatus::Ready) {
+                throw new CloneException(
+                    trans('courses.finder.clone_in.error'),
+                );
+            }
+        }
+    }
+
+    /**
+     * Clone this card.
+     *
+     * All attachments will be cloned as well (but not regular file).
+     *
+     * When cloned in a new course, all files and tags will be cloned.
+     *
+     * @param  Folder|null  $destFolder The new parent folder. Null if the card
+     * should be cloned in the same parent folder.
+     * @param  Course|null  $course The new course. Null if the card should be
+     * cloned in the same course.
+     *
+     * @throws InvalidArgumentException If both $destFolder and $destCourse are
+     * specified.
+     */
+    public function clone(
```

Comment:
It's a bit complicated since they have many state variables ($files, $success, etc) that should be given and returned / modified in all these private functions. I don't think it will make the code much more readable because of that.